arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+    // The move constructor of std::multimap is **not** noexcept.
+    // Thus we have to make a new map unique and use std::swap to
+    // transfer the content. And then move the unique_ptr into
+    // the lambda.
+    auto queue = std::make_unique<WaitForQueue>(std::move(leaderData._waitForQueue));
+
+    auto action = [promises = std::move(queue)]() mutable noexcept {
+      for (auto& [idx, promise] : *promises) {
+        // Check this to make sure that setException does not throw
+        if (!promise.isFulfilled()) {
+          promise.setException(basics::Exception(TRI_ERROR_REPLICATION_LEADER_CHANGE,
+                                                 __FILE__, __LINE__));
+        }
+      }
+    };
+
+    LOG_CTX("8696f", INFO, _logContext) << "resign";
+    leaderData._didResign = true;
+    static_assert(
+        std::is_nothrow_constructible_v<DeferredAction, std::add_rvalue_reference_t<decltype(action)>>);
+    static_assert(noexcept(std::declval<LocalFollower&&>().resign()));
+    return std::make_tuple(std::move(localFollower).resign(),
+                           DeferredAction(std::move(action)));
+  });
+}
+
+auto replicated_log::LogLeader::readReplicatedEntryByIndex(LogIndex idx) const
+    -> std::optional<PersistingLogEntry> {
+  return _guardedLeaderData.doUnderLock([&idx](auto& leaderData) -> std::optional<PersistingLogEntry> {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    if (auto entry = leaderData._inMemoryLog.getEntryByIndex(idx);
+        entry.has_value() && entry->entry().logIndex() <= leaderData._commitIndex) {
+      return entry->entry();
+    } else {
+      return std::nullopt;
+    }
+  });
+}
+
+auto replicated_log::LogLeader::getStatus() const -> LogStatus {
+  return _guardedLeaderData.doUnderLock([term = _currentTerm](auto& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    LeaderStatus status;
+    status.local = leaderData.getLocalStatistics();
+    status.term = term;
+    for (FollowerInfo const& f : leaderData._follower) {
+      status.follower[f._impl->getParticipantId()] = {
+          LogStatistics{f.lastAckedEntry, f.lastAckedCommitIndex}, f.lastErrorReason,
+          std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(f._lastRequestLatency)
+              .count()};
+    }
+    return LogStatus{std::move(status)};
+  });
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, bool waitForSync) -> LogIndex {
+  auto index = insert(std::move(payload), waitForSync, doNotTriggerAsyncReplication);
+  triggerAsyncReplication();
+  return index;
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, [[maybe_unused]] bool waitForSync,
+                                       DoNotTriggerAsyncReplication) -> LogIndex {
+  // TODO Handle waitForSync!
```

Comment:
Uh, should we add support for this in this PR already?

Reply:
I would be getting this PR through first. There will be follow up PR that will need this feature and thus it will be implemented sooner or later anyways.